snippets/components/video-render.js

```diff
@@ -0,0 +1,147 @@
+class LazyVideoComponent extends HTMLElement {
+  constructor() {
+    super();
+
+    const width = this.getAttribute("width");
+    const height = this.getAttribute("height");
+
+    const aspectRatio = this.getAspectRatio(width, height);
+
+    this.videoElement = document.createElement("template");
+    this.videoElement.innerHTML = `
+        <style>
+          video {
+            max-width: 100%;
+            aspect-ratio: ${aspectRatio};
+          }
+        </style>
+        
+        <video muted autoplay></video>  
+      `;
+
+    if (!this.shadowRoot) {
+      this.attachShadow({ mode: "open" });
+      this.shadowRoot.append(this.videoElement.content.cloneNode(true));
+    }
+
+    this._isLoaded = false;
+    this.observer = null;
+    this.update();
+  }
+
+  connectedCallback() {
+    this.setObserver();
+  }
+
+  disconnectedCallback() {
+    this.observer?.disconnect();
+  }
+
+  get isLoaded() {
+    return this._isLoaded;
+  }
+
+  set isLoaded(value) {
+    this._isLoaded = value;
+  }
+
+  setObserver() {
+    this.observer = new IntersectionObserver((entries) => {
+      const video = this.shadowRoot.querySelector("video");
+
+      entries.forEach((entry) => {
+        if (entry.isIntersecting) {
+          this.handleVideoInView(video);
+        } else {
+          this.handleVideoOutOfView(video);
+        }
+      });
+    });
+
+    this.observer.observe(this);
+  }
+
+  handleVideoInView(video) {
+    if (!this.isLoaded) {
+      this.loadVideo(video);
+      this.observer.unobserve(video);
+    } else {
+      video.play();
+    }
+  }
+
+  handleVideoOutOfView(video) {
+    if (this.isLoaded) {
+      video.pause();
+    }
+  }
+
+  loadVideo(video) {
+    Array.from(video.children).forEach((child) => {
+      if (child.tagName === "source".toUpperCase()) {
+        child.src = child.getAttribute("data-src");
+      }
+    });
+
+    video.load();
+    this.isLoaded = true;
+  }
+
+  update() {
+    const videoTag = this.shadowRoot.querySelector("video");
+    const videoSources = JSON.parse(this.getAttribute("sources"));
+
+    videoSources.forEach((source) => {
+      let sourceTag = document.createElement("source");
+
+      sourceTag.setAttribute("data-src", source.url);
+      sourceTag.setAttribute("type", source.mime_type);
+
+      videoTag.append(sourceTag);
+    });
+
+    this.addVideoAttributes(videoTag);
+    this.removeComponentAttributes();
+
+    if (videoTag.getAttribute("fullwidth")) this.setFullWidth(videoTag);
+  }
+
+  addVideoAttributes(videoTag) {
+    const allAtributes = this.getAttributeNames();
+
+    allAtributes.forEach((attribute) => {
+      const value = this.getAttribute(attribute);
+      videoTag.setAttribute(attribute, value);
+    });
+
+    this.removeVideoAttributes();
```

Comment:
Why are we allowing them to be added and removed after the fact? If the goal here is to avoid having these attributes, we could do something like this.

```
const rejected = ["class", "sources", "width", "height"];
allAtributes.forEach((attribute) => {
  if (rejected.includes(attribute) return;
  const value = this.getAttribute(attribute);
  videoTag.setAttribute(attribute, value);
});
```

Let me know if I'm misunderstanding at all.